Allow setting a custom placement policy per table through yb-admin_cl…

deeps1991 opened the pull request:
…i that overrides the cluster placement policy.

vvkgopalan commented:
One case to account for is when a table is colocated / part of a tablegroup, in which case setting a custom placement policy for that table should be dis-allowed. This is because colocated tables / tables within a tablegroup are placed on a singular parent tablet. For reference on implementation, see https://github.com/yugabyte/yugabyte-db/commit/6ef74d88f53dbab0af52ba793a9aaa5a7e5799a5 and https://github.com/yugabyte/yugabyte-db/commit/9fa72cd3048bacf35bc8aab7191a702ed0a6271b.

In the case of tablegroups, it may make sense to have a separate command to enable setting this placement policy per-tablegroup (so that all the tables within the tablegroup also obey it).

deeps1991 commented on src/yb/tools/yb-admin_client.h:
```diff
@@ -189,6 +189,11 @@ class ClusterAdminClient {
+  CHECKED_STATUS ModifyTablePlacementInfo(const client::YBTableName& table_name,
+                                          const std::string& placement_info,
+                                          int replication_factor,
+                                          const std::string& optional_uuid);
```
I left it here to be future proof, but added a check to ensure that the two uuids match. Let me know if that is okay, or I can remove this in the next patch.

deeps1991 commented:
Thanks for the review Vivek and Rahul! I have addressed all the comments in the last commit. 
@rahuldesirazu I've implemented the change based on our offline slack discussion - and implemented checks to verify that
a) We set the custom table placement info only for a cluster with just live placement info set.
b) The placement uuid for the table matches the placement uuid of the cluster
c) We only set live placement info for a table.

Let me know if my understanding went incorrect somewhere, and I can fix those in a follow up patch.

rahuldesirazu commented on src/yb/integration-tests/load_balancer_placement_policy-test.cc:
```diff
@@ -0,0 +1,336 @@
+  ASSERT_NO_FATALS(cluster_verifier.CheckRowCount(
+    placement_table, ClusterVerifier::EXACTLY, rows_inserted));
+}
+
```
Not sure if its already tested, but I would add/modify a test that calls ModifyPlacementInfo with the optional placement_uuid set (3rd argument) and make sure table level overrides still work.

deeps1991 replied:
Done

rahuldesirazu reviewed (commented):
Everything looking really solid! Just a small test case to add and a nit fix.

deeps1991 commented on src/yb/client/table_alterer.h:
```diff
@@ -111,6 +114,10 @@ class YBTableAlterer {
 
   boost::optional<uint32_t> wal_retention_secs_;
 
+  master::ReplicationInfoPB replication_info_;
```
Done

deepthisrinivasan commented on src/yb/integration-tests/load_balancer_placement_policy-test.cc:
```diff
@@ -0,0 +1,395 @@
+  ASSERT_OK(yb_admin_client_->ModifyTablePlacementInfo(
+    table_name(), "c.r.z2", 1, random_placement_uuid));
+  WaitForLoadBalancer();
+
```
This is actually covered in admin-cli-test

deeps1991 commented on src/yb/integration-tests/load_balancer_placement_policy-test.cc:
```diff
@@ -0,0 +1,395 @@
+  ASSERT_EQ(counts_per_ts[4], 4);
+
+}
+
```
This is covered in first few steps of PlacementPolicyTest
